src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -0,0 +1,351 @@
+            }
+        }
+
+        // Any exception thrown in this method is left to bubble up to the extension.
+        var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+        var result = await handler.ExecuteAsync(message, document, cancellationToken)
+            .ConfigureAwait(false);
+        var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+        return responseJson;
+    }
+
+    public ValueTask UnregisterExtensionAsync(
+        string assemblyFilePath,
+        CancellationToken cancellationToken)
+    {
+        var assemblyFileName = Path.GetFileName(assemblyFilePath);
+        var assemblyFolderPath = Path.GetDirectoryName(assemblyFilePath)
+            ?? throw new InvalidOperationException($"Unable to get the directory name for {assemblyFilePath}.");
+
+        try
+        {
+            CustomMessageHandlerExtension? extension = null;
+            lock (_lockObject)
+            {
+                if (_extensions.TryGetValue(assemblyFolderPath, out extension))
+                {
+                    if (extension.Assemblies.TryGetValue(assemblyFileName, out var extensionAssembly))
+                    {
+                        extension.Assemblies.Remove(assemblyFileName);
+                        UnregisterHandlersForAssembly(extensionAssembly);
+                    }
+
+                    if (extension.Assemblies.Count == 0)
+                    {
+                        _extensions.Remove(assemblyFolderPath);
+                        foreach (var assembly in extension.Assemblies.Values)
+                        {
+                            UnregisterHandlersForAssembly(assembly);
+                        }
+                    }
+                }
+            }
+
+            extension?.AnalyzerAssemblyLoader.Dispose();
+        }
+        catch (Exception e) when (LogAndPropagate(e))
+        {
+            // unreachable
+        }
+
+        return ValueTask.CompletedTask;
+
+        bool LogAndPropagate(Exception e)
+        {
+            Logger.Log(
+                FunctionId.CustomMessageHandlerService_UnloadCustomMessageHandlerAsync,
+                $"Error unregistering {assemblyFilePath}: {e}",
+                LogLevel.Error);
+            return false;
+        }
+
+        void UnregisterHandlersForAssembly(CustomMessageHandlerAssembly? assembly)
+        {
+            if (assembly.HasValue)
+            {
+                foreach (var handler in assembly.Value.WorkspaceMessageHandlers)
+                {
+                    _handlers.Remove(handler);
+                }
+
+                foreach (var documentHandler in assembly.Value.DocumentMessageHandlers)
+                {
+                    _documentHandlers.Remove(documentHandler);
+                }
+            }
+        }
+    }
+
+    public ValueTask ResetAsync(CancellationToken cancellationToken)
+    {
+        Clear();
+        return ValueTask.CompletedTask;
```

Comment:
why even have an async method here if it's neither cancellable, nor ever does anything asynchronous?

Reply:
This was a brokered service at some point, so every method was async. Now that it's not anymore, I can make it a sync method.
All other methods also complete synchronously for now (since we use locks) but I will wait to drop the Async pattern for now, in case we decide to do something different with the locking